crates/matrix-sdk/src/client/mod.rs

```diff
@@ -1510,16 +1544,41 @@ impl Client {
     /// let msc_x = unstable_features.get("msc_x").unwrap_or(&false);
     /// # anyhow::Ok(()) };
     /// ```
-    pub async fn unstable_features(&self) -> HttpResult<&BTreeMap<String, bool>> {
-        let unstable_features = self
-            .inner
-            .unstable_features
-            .get_or_try_init(|| self.request_unstable_features())
-            .await?;
+    pub async fn unstable_features(&self) -> HttpResult<BTreeMap<String, bool>> {
+        if let Some(unstable_features) = self.inner.unstable_features.read().await.as_ref() {
+            return Ok(unstable_features.clone());
+        }
+
+        let mut guard = self.inner.unstable_features.write().await;
+        if let Some(unstable_features) = guard.as_ref() {
+            return Ok(unstable_features.clone());
+        }
+
+        let (versions, unstable_features) = self.load_or_fetch_server_capabilities().await?;
+
+        *guard = Some(unstable_features.clone());
+        drop(guard);
+
+        // Fill the versions, while we're at it.
+        *self.inner.server_versions.write().await = Some(versions);
 
         Ok(unstable_features)
     }
 
+    /// Empty the server version and unstable features cache.
+    ///
+    /// Since the SDK caches server capabilities (versions and unstable
+    /// features), it's possible to have a stale entry in the cache. This
+    /// functions makes it possible to force reset it.
+    pub async fn reset_server_capabilities(&self) -> Result<()> {
+        // Empty the in-memory caches.
+        *self.inner.server_versions.write().await = None;
```

Comment:
I think that this is fine, as you eagerly call drop in the previous methods, so there's no deadlock. But modifying any of the above methods could easily introduce one. I don't see a reason for this not being behind the same lock.